### ISSUE TITLE
Deadlock on Offline Shadow Update

### ISSUE THREAD

**Issue by kc9zyz** [aws/aws-iot-device-sdk-python#52]
# Preconditions:
1. Two Device shadows are associated (as in a gateway scenerio), but no update has been sent so the `update/accepted` and `update/rejected` topics are not yet subscribed to.

# Process to Re-Create:
1. Remove connectivity from the library 
1. Attempt to update one of the shadows, receive a subscribeTimoutException
1. Attempt to update the second shadow

# Result
The call to update the second shadow never returns. The offending call is [here](https://github.com/aws/aws-iot-device-sdk-python/blob/cb67058517681d2ffddc8064b1439988c78ee822/AWSIoTPythonSDK/core/shadow/shadowManager.py#L68):    
```
def basicShadowSubscribe(self, srcShadowName, srcShadowAction, srcCallback):
        self._shadowSubUnsubOperationLock.acquire()
        currentShadowAction = _shadowAction(srcShadowName, srcShadowAction)
        if currentShadowAction.isDelta:
            self._mqttCoreHandler.subscribe(currentShadowAction.getTopicDelta(), 0, srcCallback)
        else:
            self._mqttCoreHandler.subscribe(currentShadowAction.getTopicAccept(), 0, srcCallback)
            self._mqttCoreHandler.subscribe(currentShadowAction.getTopicReject(), 0, srcCallback)
        time.sleep(2)
        self._shadowSubUnsubOperationLock.release()
```

On line 68 a lock is acquired. Notice that on lines 71, 73, and 74, a subscribeTimeoutException may be raised (as happened in the second step to recreate), jumping past the lock release call on line 76. Any cases of locks being acquired/released should be using context managers or exception handling to ensure that the lock is released if an operation fails.

**Comment by liuszeng** [comment]
Hi @kc9zyz ,

Thank you very much for pointing this out! We will incorporate the bug fix into the next release.

Thanks,
Liusu

**Comment by liuszeng** [comment]
Addressed in release [v1.2.0](https://github.com/aws/aws-iot-device-sdk-python/releases/tag/v1.2.0).

**Comment by kc9zyz** [comment]
@liuszeng Thanks for the update.
